java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/protocol/LspTemplateUI.java

```diff
@@ -0,0 +1,215 @@
+
+    @NbBundle.Messages({
+        "CTL_TemplateUI_SelectGroup=Select Group of Objects",
+        "CTL_TemplateUI_SelectTemplate=Select Object Template",
+        "CTL_TemplateUI_SelectTarget=Where to put the object?",
+        "CTL_TemplateUI_SelectName=Name of the object?",
+    })
+    private static CompletableFuture<Object> displayUI(DataFolder templates, NbCodeLanguageClient client, ExecuteCommandParams params) {
+        final FileObject[] group = { null };
+        final DataObject[] source = { null };
+        final DataFolder[] target = findTargetFolder(params);
+
+        final List<QuickPickItem> categories = quickPickTemplates(templates);
+        final CompletionStage<List<QuickPickItem>> pickGroup = client.showQuickPick(new ShowQuickPickParams(Bundle.CTL_TemplateUI_SelectGroup(), false, categories));
+        final CompletionStage<List<QuickPickItem>> pickProject = pickGroup.thenCompose(selectedGroups -> {
+            group[0] = templates.getPrimaryFile().getFileObject(singleSelection(selectedGroups));
+            List<QuickPickItem> projectTypes = quickPickTemplates(DataFolder.findFolder(group[0]));
+            return client.showQuickPick(new ShowQuickPickParams(Bundle.CTL_TemplateUI_SelectTemplate(), false, projectTypes));
+        });
+        final CompletionStage<DataObject> findTemplate = pickProject.thenApply((selectedTemplates) -> {
+            try {
+                final String templateName = singleSelection(selectedTemplates);
+                final FileObject templateFo = group[0].getFileObject(templateName);
+                source[0] = DataObject.find(templateFo);
+            } catch (DataObjectNotFoundException ex) {
+                throw raise(RuntimeException.class, ex);
+            }
+            return source[0];
+        });
+        final CompletionStage<DataFolder> findTarget;
+        if (target[0] == null) {
+            findTarget = findTemplate.thenCompose(any -> client.workspaceFolders()).thenCompose(folders -> {
+                return client.showInputBox(new ShowInputBoxParams(Bundle.CTL_TemplateUI_SelectTarget(), findWorkspaceRoot(folders)));
+            }).thenApply((path) -> {
+                target[0] = DataFolder.findFolder(FileUtil.toFileObject(new File(path)));
+                return target[0];
+            });
+        } else {
+            findTarget = findTemplate.thenApply(any -> target[0]);
+        }
+        CompletionStage<String> findTargetName = findTarget.thenCompose((t) -> {
+            return client.showInputBox(new ShowInputBoxParams(Bundle.CTL_TemplateUI_SelectName(), source[0].getName()));
+        }).thenApply((nameWithExtension) -> {
+            String templateExtension = source[0].getPrimaryFile().getExt();
+            return removeExtensionFromFileName(nameWithExtension, templateExtension);
+        });
+        return findTargetName.thenApply((name) -> {
+            try {
+                DataObject newPrj = source[0].createFromTemplate(target[0], name);
```

Comment:
`name` can potentially contain invalid user input (i.e. '/' in the name) - are error messages from `createForTemplate` reasonable ?

Reply:
Using `/` in the name creates a subdirectory and uses wrong name for class name. No error is reported.